In [1]:
import os
import sys
spark_path = os.environ['SPARK_HOME']
sys.path.append(spark_path + "/bin")
sys.path.append(spark_path + "/python")
sys.path.append(spark_path + "/python/pyspark/")
sys.path.append(spark_path + "/python/lib")
sys.path.append(spark_path + "/python/lib/pyspark.zip")
sys.path.append(spark_path + "/python/lib/py4j-0.10.9-src.zip")

import findspark
findspark.init()
import pyspark

In [2]:
number_cores = 6
memory_gb = 16
conf = (pyspark.SparkConf().setMaster('local[{}]'.format(number_cores)).set('spark.driver.memory', '{}g'.format(memory_gb)))
sc = pyspark.SparkContext(conf=conf)

In [3]:
!dir /users/trush/CSC496/Labs/Lab3

data  Lab3.ipynb


## How do you parse JSON?

In [4]:
import json

def parseJSON(t):
    res = json.loads(t)
    return res

test = '{"user_id":"ntlvfPzc8eglqvk92iDIAw","name":"Rafael","review_count":553,"yelping_since":"2007-07-06 03:27:11","useful":628,"funny":225,"cool":227,"elite":"","friends":"oeMvJh94PiGQnx_6GlndPQ, wm1z1PaJKvHgSDRKfwhfDg, IkRib6Xs91PPW7pon7VVig, A8Aq8f0-XvLBcyMk2GJdJQ, eEZM1kogR7eL4GOBZyPvBA, e1o1LN7ez5ckCpQeAab4iw, _HrJVzFaRFUhPva8cwBjpQ, pZeGZGzX-ROT_D5lam5uNg, 0S6EI51ej5J7dgYz3-O0lA, woDt8raW-AorxQM_tIE2eA, hWUnSE5gKXNe7bDc8uAG9A, c_3LDSO2RHwZ94_Q6j_O7w, -uv1wDiaplY6eXXS0VwQiA, QFjqxXn3acDC7hckFGUKMg, ErOqapICmHPTN8YobZIcfQ, mJLRvqLOKhqEdkgt9iEaCQ, VKX7jlScJSA-ja5hYRw12Q, ijIC9w5PRcj3dWVlanjZeg, CIZGlEw-Bp0rmkP8M6yQ9Q, OC6fT5WZ8EU7tEVJ3bzPBQ, UZSDGTDpycDzrlfUlyw2dQ, deL6e_z9xqZTIODKqnvRXQ, 5mG2ENw2PylIWElqHSMGqg, Uh5Kug2fvDd51RYmsNZkGg, 4dI4uoShugD9z84fYupelQ, EQpFHqGT9Tk6YSwORTtwpg, o4EGL2-ICGmRJzJ3GxB-vw, s8gK7sdVzJcYKcPv2dkZXw, vOYVZgb_GVe-kdtjQwSUHw, wBbjgHsrKr7BsPBrQwJf2w, p59u2EC_qcmCmLeX1jCi5Q, VSAZI1eHDrOPRWMK4Q2DIQ, efMfeI_dkhpeGykaRJqxfQ, x6qYcQ8_i0mMDzSLsFCbZg, K_zSmtNGw1fu-vmxyTVfCQ, 5IM6YPQCK-NABkXmHhlRGQ, U_w8ZMD26vnkeeS1sD7s4Q, AbfS_oXF8H6HJb5jFqhrLw, hbcjX4_D4KIfonNnwrH-cg, UKf66_MPz0zHCP70mF6p1g, hK2gYbxZRTqcqlSiQQcrtQ, 2Q45w_Twx_T9dXqlE16xtQ, BwRn8qcKSeA77HLaOTbfiQ, jouOn4VS_DtFPtMR2w8VDA, ESteyJabbfvqas6CEDs3pQ","fans":14,"average_stars":3.57,"compliment_hot":3,"compliment_more":2,"compliment_profile":1,"compliment_cute":0,"compliment_list":1,"compliment_note":11,"compliment_plain":15,"compliment_cool":22,"compliment_funny":22,"compliment_writer":10,"compliment_photos":0}'
parseJSON(test)

{'user_id': 'ntlvfPzc8eglqvk92iDIAw',
 'name': 'Rafael',
 'review_count': 553,
 'yelping_since': '2007-07-06 03:27:11',
 'useful': 628,
 'funny': 225,
 'cool': 227,
 'elite': '',
 'friends': 'oeMvJh94PiGQnx_6GlndPQ, wm1z1PaJKvHgSDRKfwhfDg, IkRib6Xs91PPW7pon7VVig, A8Aq8f0-XvLBcyMk2GJdJQ, eEZM1kogR7eL4GOBZyPvBA, e1o1LN7ez5ckCpQeAab4iw, _HrJVzFaRFUhPva8cwBjpQ, pZeGZGzX-ROT_D5lam5uNg, 0S6EI51ej5J7dgYz3-O0lA, woDt8raW-AorxQM_tIE2eA, hWUnSE5gKXNe7bDc8uAG9A, c_3LDSO2RHwZ94_Q6j_O7w, -uv1wDiaplY6eXXS0VwQiA, QFjqxXn3acDC7hckFGUKMg, ErOqapICmHPTN8YobZIcfQ, mJLRvqLOKhqEdkgt9iEaCQ, VKX7jlScJSA-ja5hYRw12Q, ijIC9w5PRcj3dWVlanjZeg, CIZGlEw-Bp0rmkP8M6yQ9Q, OC6fT5WZ8EU7tEVJ3bzPBQ, UZSDGTDpycDzrlfUlyw2dQ, deL6e_z9xqZTIODKqnvRXQ, 5mG2ENw2PylIWElqHSMGqg, Uh5Kug2fvDd51RYmsNZkGg, 4dI4uoShugD9z84fYupelQ, EQpFHqGT9Tk6YSwORTtwpg, o4EGL2-ICGmRJzJ3GxB-vw, s8gK7sdVzJcYKcPv2dkZXw, vOYVZgb_GVe-kdtjQwSUHw, wBbjgHsrKr7BsPBrQwJf2w, p59u2EC_qcmCmLeX1jCi5Q, VSAZI1eHDrOPRWMK4Q2DIQ, efMfeI_dkhpeGykaRJqxfQ, x6qYcQ8_i0mMDzSL

In [5]:
graph_data = sc.textFile("/users/trush/CSC496/Labs/Lab3/data/yelp_small.json")
graph_data.take(1)

['{"user_id":"ntlvfPzc8eglqvk92iDIAw","name":"Rafael","review_count":553,"yelping_since":"2007-07-06 03:27:11","useful":628,"funny":225,"cool":227,"elite":"","friends":"oeMvJh94PiGQnx_6GlndPQ, wm1z1PaJKvHgSDRKfwhfDg, IkRib6Xs91PPW7pon7VVig, A8Aq8f0-XvLBcyMk2GJdJQ, eEZM1kogR7eL4GOBZyPvBA, e1o1LN7ez5ckCpQeAab4iw, _HrJVzFaRFUhPva8cwBjpQ, pZeGZGzX-ROT_D5lam5uNg, 0S6EI51ej5J7dgYz3-O0lA, woDt8raW-AorxQM_tIE2eA, hWUnSE5gKXNe7bDc8uAG9A, c_3LDSO2RHwZ94_Q6j_O7w, -uv1wDiaplY6eXXS0VwQiA, QFjqxXn3acDC7hckFGUKMg, ErOqapICmHPTN8YobZIcfQ, mJLRvqLOKhqEdkgt9iEaCQ, VKX7jlScJSA-ja5hYRw12Q, ijIC9w5PRcj3dWVlanjZeg, CIZGlEw-Bp0rmkP8M6yQ9Q, OC6fT5WZ8EU7tEVJ3bzPBQ, UZSDGTDpycDzrlfUlyw2dQ, deL6e_z9xqZTIODKqnvRXQ, 5mG2ENw2PylIWElqHSMGqg, Uh5Kug2fvDd51RYmsNZkGg, 4dI4uoShugD9z84fYupelQ, EQpFHqGT9Tk6YSwORTtwpg, o4EGL2-ICGmRJzJ3GxB-vw, s8gK7sdVzJcYKcPv2dkZXw, vOYVZgb_GVe-kdtjQwSUHw, wBbjgHsrKr7BsPBrQwJf2w, p59u2EC_qcmCmLeX1jCi5Q, VSAZI1eHDrOPRWMK4Q2DIQ, efMfeI_dkhpeGykaRJqxfQ, x6qYcQ8_i0mMDzSLsFCbZg, K_zSmtNGw1fu-vm

## Key points:
Always try and make your data smaller. Cache often if possible.

### Hands-on:
1. How do you apply pagerank on num_num?
2. How do you link the answers back to the actual page of hollies.dat?

In [6]:
from pyspark.sql import SQLContext
sqlContext = SQLContext(sc)

In [7]:
json_df = sqlContext.read.json("/users/trush/CSC496/Labs/Lab3/data/yelp_small.json")

In [8]:
%%time
json_df.count()

CPU times: user 469 µs, sys: 2.25 ms, total: 2.72 ms
Wall time: 923 ms


1000

In [9]:
%%time
json_df.count()

CPU times: user 2.49 ms, sys: 1.19 ms, total: 3.69 ms
Wall time: 238 ms


1000

In [10]:
json_df.printSchema()

root
 |-- average_stars: double (nullable = true)
 |-- compliment_cool: long (nullable = true)
 |-- compliment_cute: long (nullable = true)
 |-- compliment_funny: long (nullable = true)
 |-- compliment_hot: long (nullable = true)
 |-- compliment_list: long (nullable = true)
 |-- compliment_more: long (nullable = true)
 |-- compliment_note: long (nullable = true)
 |-- compliment_photos: long (nullable = true)
 |-- compliment_plain: long (nullable = true)
 |-- compliment_profile: long (nullable = true)
 |-- compliment_writer: long (nullable = true)
 |-- cool: long (nullable = true)
 |-- elite: string (nullable = true)
 |-- fans: long (nullable = true)
 |-- friends: string (nullable = true)
 |-- funny: long (nullable = true)
 |-- name: string (nullable = true)
 |-- review_count: long (nullable = true)
 |-- useful: long (nullable = true)
 |-- user_id: string (nullable = true)
 |-- yelping_since: string (nullable = true)



In [11]:
json_df.registerTempTable("tbl_json")

## 1. How many unique users are there in the dataset?

In [12]:
def parseJSONUser(x):
    res = json.loads(x)
    return res['user_id']

In [13]:
# map with just user_id .distinct().count()
users = graph_data.map(lambda x: parseJSONUser(x))
users.take(5)
users.distinct().count()
#graph_data.countByKey()

1000

There are 1000 unique users in the small dataset.

## 2. Identify the user(s) with the highest number of reviews. What is(are) the average/median stars for these user(s)?

In [14]:
def parseJSONReviews(x):
    res = json.loads(x)
    return (res['user_id'], res['name'], res['average_stars'], int(res['review_count']))

In [15]:
# map with user_id and review_count
user_data = graph_data.map(lambda x: parseJSON(x))
user_data.take(1)
# ** HOW MANY USERS TO AVERAGE (TOP 100?)? ** ** DO YOU WANT USER_ID OR NAME? **
highest_num_reviews = graph_data.map(lambda x: parseJSONReviews(x)).sortBy(lambda x: (x[3]), ascending = False)
top10 = highest_num_reviews.take(10)
top10

[('8k3aO-mPeyhbR5HUucA5aA', 'Victor', 3.28, 14455),
 ('hWDybu_KvYLSdEFzGrniTw', 'Bruce', 3.64, 12487),
 ('Hi10sGSZNxQH3NLyWSZ1oA', 'Fox', 3.8, 11112),
 ('XYSDrIef7g4Gmp3lNFVO6A', 'Neal', 3.96, 4828),
 ('CQUDh80m48xnzUkx-X5NAw', 'David', 3.9, 4181),
 ('YttDgOC9AlM4HcAlDsbB2A', 'Phil', 3.58, 3379),
 ('ITa3vh5ERI90G_WP4SmGUQ', 'Peter', 3.99, 3079),
 ('PkNs-Rw6MHQfu6j5pBZMMw', 'Maria', 3.67, 2978),
 ('ZIOCmdFaMIF56FR-nWr_2A', 'Randy', 3.77, 2922),
 ('Ryxj0u0AW3mRsRypdYli2A', 'Arnel', 4.29, 2642)]

In [16]:
# Average stars for users with top 10 highest number of reviews
def avg_top10(x):
    total = 0
    for i in range(10):
        total = total + top10[i][2]
    return total/10
# Median stars for users with top 10 highest number of reviews
import statistics
def med_top10(x):
    x.sort(key = lambda x: float(x[2]))
    return x[len(x)//2][2]

In [17]:
print("Average: " + str(avg_top10(top10)))
print("Median: " + str(med_top10(top10)))

Average: 3.7880000000000003
Median: 3.8


The top 10 users with the highest number of reviews are:
1. 8k3aO-mPeyhbR5HUucA5aA (Victor) with 14455 reviews
2. hWDybu_KvYLSdEFzGrniTw (Bruce) with 12487 reviews
3. Hi10sGSZNxQH3NLyWSZ1oA (Fox) with 11112 reviews
4. XYSDrIef7g4Gmp3lNFVO6A (Neal) with 4828 reviews
5. CQUDh80m48xnzUkx-X5NAw (David) with 4181 reviews
6. YttDgOC9AlM4HcAlDsbB2A (Phil) with 3379 reviews
7. ITa3vh5ERI90G_WP4SmGUQ (Peter) with 3079 reviews
8. PkNs-Rw6MHQfu6j5pBZMMw (Maria) with 2978 reviews
9. ZIOCmdFaMIF56FR-nWr_2A (Randy) with 2922 reviews
10. Ryxj0u0AW3mRsRypdYli2A (Arnel) with 2642 reviews

The average stars for the top 10 users is about 3.79 stars. The median stars for the top 10 users is 3.8 stars.

## 3. Identify several (two or more) interesting user attributes and think of and carry out one interesting analysis. Interesting is relative, and you should use Markdown cell to describe why/how these are interesting!

In [18]:
# friends, fans, elite, average_stars
# numFriends relative to numFans
# Are numFriends and/or numFans and/or average_stars correlated to elite status?
#user_data.take(2)

In [19]:
def parseJSONStars(x):
    res = json.loads(x)
    return (res['user_id'], res['average_stars'], len(res['elite']), len(res['friends']), res['fans'])

In [20]:
stars = graph_data.map(lambda x: parseJSONStars(x)).sortBy(lambda x: (x[3]), ascending = False)
stars.take(10)

[('ZIOCmdFaMIF56FR-nWr_2A', 3.77, 39, 250342, 1023),
 ('IU86PZPgTDCFwJEuAg2j7g', 4.13, 49, 208126, 1313),
 ('djxnI8Ux8ZYQJhiOQkrRhA', 4.12, 49, 206158, 1614),
 ('fgwI3rYHOv1ipfVfCSx7pg', 4.31, 49, 191734, 2113),
 ('VHdY6oG2JPVNjihWhOooAQ', 4.19, 64, 170110, 2140),
 ('IDVFG1pNSHIHoVuoLuZpcQ', 3.88, 14, 166918, 1233),
 ('YttDgOC9AlM4HcAlDsbB2A', 3.58, 39, 164014, 756),
 ('dIIKEfOgo0KqUfGQvGikPg', 4.07, 59, 150142, 971),
 ('ITa3vh5ERI90G_WP4SmGUQ', 3.99, 54, 149206, 2280),
 ('yh8_9Ix1wp9dekvyOZbilQ', 4.22, 49, 145126, 1040)]

## 4. Identify the top-ten most influential users from the Yelp dataset. Justify the attribute(s) that you pick to measure the level of influence!  

In [21]:
# PageRank: add weight for elite attribute
# Use numFriends and/or numFans according and/or average_stars to findings from 3

In [22]:
def parseJSONFriends(x):
    res = json.loads(x)
    return (res['user_id'], res['friends'].split(", "))

In [23]:
links = graph_data.map(lambda x: (parseJSONFriends(x))) \
        .mapValues(list)
links.take(1)

[('ntlvfPzc8eglqvk92iDIAw',
  ['oeMvJh94PiGQnx_6GlndPQ',
   'wm1z1PaJKvHgSDRKfwhfDg',
   'IkRib6Xs91PPW7pon7VVig',
   'A8Aq8f0-XvLBcyMk2GJdJQ',
   'eEZM1kogR7eL4GOBZyPvBA',
   'e1o1LN7ez5ckCpQeAab4iw',
   '_HrJVzFaRFUhPva8cwBjpQ',
   'pZeGZGzX-ROT_D5lam5uNg',
   '0S6EI51ej5J7dgYz3-O0lA',
   'woDt8raW-AorxQM_tIE2eA',
   'hWUnSE5gKXNe7bDc8uAG9A',
   'c_3LDSO2RHwZ94_Q6j_O7w',
   '-uv1wDiaplY6eXXS0VwQiA',
   'QFjqxXn3acDC7hckFGUKMg',
   'ErOqapICmHPTN8YobZIcfQ',
   'mJLRvqLOKhqEdkgt9iEaCQ',
   'VKX7jlScJSA-ja5hYRw12Q',
   'ijIC9w5PRcj3dWVlanjZeg',
   'CIZGlEw-Bp0rmkP8M6yQ9Q',
   'OC6fT5WZ8EU7tEVJ3bzPBQ',
   'UZSDGTDpycDzrlfUlyw2dQ',
   'deL6e_z9xqZTIODKqnvRXQ',
   '5mG2ENw2PylIWElqHSMGqg',
   'Uh5Kug2fvDd51RYmsNZkGg',
   '4dI4uoShugD9z84fYupelQ',
   'EQpFHqGT9Tk6YSwORTtwpg',
   'o4EGL2-ICGmRJzJ3GxB-vw',
   's8gK7sdVzJcYKcPv2dkZXw',
   'vOYVZgb_GVe-kdtjQwSUHw',
   'wBbjgHsrKr7BsPBrQwJf2w',
   'p59u2EC_qcmCmLeX1jCi5Q',
   'VSAZI1eHDrOPRWMK4Q2DIQ',
   'efMfeI_dkhpeGykaRJqxfQ',
   'x6qYcQ8_i0m

In [24]:
N = links.count()
ranks = links.map(lambda line: (line[0], 1/N))
ranks.take(1)
#links.count()

[('ntlvfPzc8eglqvk92iDIAw', 0.001)]

In [38]:
votes = ranks.join(links)
votes.take(1)

[('G5LXEaxhQMF_BVjNHkVr7g',
  (6.372625885950604e-08,
   ['CVSX082_G87suX8-zGuXMQ',
    '302NfFT_rIej5TTKtsnN4w',
    'mGMrvBN9nXq6uipetkROeQ',
    'Y20JOmRsD92VNo-XozNypA',
    'aQKUKYiIrj24FpElG7oQEw',
    'G9zQFp7TXeryx9K_Ri8lVA',
    'E34UQ8qK8Zyz7oCcJig0Nw',
    'ToIFzSAeEi5UFIoTmy2qLA',
    'zbhm2W7WnRXoOlx70MN53w',
    'Sc7T4XlLnqhvk6bk0I6iFA',
    'VM5g9IJmaL3JrRQ6sHYDiw',
    '55UFq4i1nTFtB3qlSbTxyw',
    'PvM3RdhfkVV1t0SzwVQEnA',
    'amw1u4EwW40pi_kpl5-x2A',
    '5pkKmvYlcxQNag1ym6gaEg',
    'OqH85SAF0oLijVY_SeAk8A',
    'ckQyi6cnOyLE1LWrfn-qww',
    'EIMmVSasYi9VYrq4IolYnA',
    'mqBvAXMHQnup85cepxPXZg',
    'L-RyP9gsSWXMv8eFsTWNRw',
    'PAbX2enaEQKFQoPgwB3-_A',
    'QgOPMVwxVmEBwGkbxQdhgA',
    '3XtMlMcY4k1Rop3aCUrn6Q',
    'zxF7vwh32ccSN3CbCMlalw',
    'CIJcnwH8WcjdO4s4ESPXNg',
    '7HGh5yCWSgYA6PdqU9BoGA',
    'x2J2d9iiRiAq3MBryWequQ',
    'o-8mFe4K3vyxT1i806Px9w',
    'U-_d1qMpvHgX8s4v0zokYg',
    'SXEaqdvMiYs2LNxeIbM51w',
    'l_v1Mx8vTJtvh4qn0EtpFA',
    'rqv9mIc9pj9

In [26]:
def calculateVotes(t):
    res = []
    for item in t[1][1]:
        count = len(t[1][1])
        res.append((item, t[1][0]/count))
    return res
calculateVotes(('ntlvfPzc8eglqvk92iDIAw',
  (0.001,
   ['oeMvJh94PiGQnx_6GlndPQ',
    'wm1z1PaJKvHgSDRKfwhfDg',
    'IkRib6Xs91PPW7pon7VVig',
    'A8Aq8f0-XvLBcyMk2GJdJQ',
    'eEZM1kogR7eL4GOBZyPvBA',
    'e1o1LN7ez5ckCpQeAab4iw',
    '_HrJVzFaRFUhPva8cwBjpQ',
    'pZeGZGzX-ROT_D5lam5uNg',
    '0S6EI51ej5J7dgYz3-O0lA',
    'woDt8raW-AorxQM_tIE2eA',
    'hWUnSE5gKXNe7bDc8uAG9A',
    'c_3LDSO2RHwZ94_Q6j_O7w',
    '-uv1wDiaplY6eXXS0VwQiA',
    'QFjqxXn3acDC7hckFGUKMg',
    'ErOqapICmHPTN8YobZIcfQ',
    'mJLRvqLOKhqEdkgt9iEaCQ',
    'VKX7jlScJSA-ja5hYRw12Q',
    'ijIC9w5PRcj3dWVlanjZeg',
    'CIZGlEw-Bp0rmkP8M6yQ9Q',
    'OC6fT5WZ8EU7tEVJ3bzPBQ',
    'UZSDGTDpycDzrlfUlyw2dQ',
    'deL6e_z9xqZTIODKqnvRXQ',
    '5mG2ENw2PylIWElqHSMGqg',
    'Uh5Kug2fvDd51RYmsNZkGg',
    '4dI4uoShugD9z84fYupelQ',
    'EQpFHqGT9Tk6YSwORTtwpg',
    'o4EGL2-ICGmRJzJ3GxB-vw',
    's8gK7sdVzJcYKcPv2dkZXw',
    'vOYVZgb_GVe-kdtjQwSUHw',
    'wBbjgHsrKr7BsPBrQwJf2w',
    'p59u2EC_qcmCmLeX1jCi5Q',
    'VSAZI1eHDrOPRWMK4Q2DIQ',
    'efMfeI_dkhpeGykaRJqxfQ',
    'x6qYcQ8_i0mMDzSLsFCbZg',
    'K_zSmtNGw1fu-vmxyTVfCQ',
    '5IM6YPQCK-NABkXmHhlRGQ',
    'U_w8ZMD26vnkeeS1sD7s4Q',
    'AbfS_oXF8H6HJb5jFqhrLw',
    'hbcjX4_D4KIfonNnwrH-cg',
    'UKf66_MPz0zHCP70mF6p1g',
    'hK2gYbxZRTqcqlSiQQcrtQ',
    '2Q45w_Twx_T9dXqlE16xtQ',
    'BwRn8qcKSeA77HLaOTbfiQ',
    'jouOn4VS_DtFPtMR2w8VDA',
    'ESteyJabbfvqas6CEDs3pQ'])))

[('oeMvJh94PiGQnx_6GlndPQ', 2.2222222222222223e-05),
 ('wm1z1PaJKvHgSDRKfwhfDg', 2.2222222222222223e-05),
 ('IkRib6Xs91PPW7pon7VVig', 2.2222222222222223e-05),
 ('A8Aq8f0-XvLBcyMk2GJdJQ', 2.2222222222222223e-05),
 ('eEZM1kogR7eL4GOBZyPvBA', 2.2222222222222223e-05),
 ('e1o1LN7ez5ckCpQeAab4iw', 2.2222222222222223e-05),
 ('_HrJVzFaRFUhPva8cwBjpQ', 2.2222222222222223e-05),
 ('pZeGZGzX-ROT_D5lam5uNg', 2.2222222222222223e-05),
 ('0S6EI51ej5J7dgYz3-O0lA', 2.2222222222222223e-05),
 ('woDt8raW-AorxQM_tIE2eA', 2.2222222222222223e-05),
 ('hWUnSE5gKXNe7bDc8uAG9A', 2.2222222222222223e-05),
 ('c_3LDSO2RHwZ94_Q6j_O7w', 2.2222222222222223e-05),
 ('-uv1wDiaplY6eXXS0VwQiA', 2.2222222222222223e-05),
 ('QFjqxXn3acDC7hckFGUKMg', 2.2222222222222223e-05),
 ('ErOqapICmHPTN8YobZIcfQ', 2.2222222222222223e-05),
 ('mJLRvqLOKhqEdkgt9iEaCQ', 2.2222222222222223e-05),
 ('VKX7jlScJSA-ja5hYRw12Q', 2.2222222222222223e-05),
 ('ijIC9w5PRcj3dWVlanjZeg', 2.2222222222222223e-05),
 ('CIZGlEw-Bp0rmkP8M6yQ9Q', 2.2222222222222223

In [27]:
votes = ranks.join(links) \
        .flatMap(calculateVotes)
votes.collect()

[('oeMvJh94PiGQnx_6GlndPQ', 2.2222222222222223e-05),
 ('wm1z1PaJKvHgSDRKfwhfDg', 2.2222222222222223e-05),
 ('IkRib6Xs91PPW7pon7VVig', 2.2222222222222223e-05),
 ('A8Aq8f0-XvLBcyMk2GJdJQ', 2.2222222222222223e-05),
 ('eEZM1kogR7eL4GOBZyPvBA', 2.2222222222222223e-05),
 ('e1o1LN7ez5ckCpQeAab4iw', 2.2222222222222223e-05),
 ('_HrJVzFaRFUhPva8cwBjpQ', 2.2222222222222223e-05),
 ('pZeGZGzX-ROT_D5lam5uNg', 2.2222222222222223e-05),
 ('0S6EI51ej5J7dgYz3-O0lA', 2.2222222222222223e-05),
 ('woDt8raW-AorxQM_tIE2eA', 2.2222222222222223e-05),
 ('hWUnSE5gKXNe7bDc8uAG9A', 2.2222222222222223e-05),
 ('c_3LDSO2RHwZ94_Q6j_O7w', 2.2222222222222223e-05),
 ('-uv1wDiaplY6eXXS0VwQiA', 2.2222222222222223e-05),
 ('QFjqxXn3acDC7hckFGUKMg', 2.2222222222222223e-05),
 ('ErOqapICmHPTN8YobZIcfQ', 2.2222222222222223e-05),
 ('mJLRvqLOKhqEdkgt9iEaCQ', 2.2222222222222223e-05),
 ('VKX7jlScJSA-ja5hYRw12Q', 2.2222222222222223e-05),
 ('ijIC9w5PRcj3dWVlanjZeg', 2.2222222222222223e-05),
 ('CIZGlEw-Bp0rmkP8M6yQ9Q', 2.2222222222222223

In [28]:
ranks.collect()

[('ntlvfPzc8eglqvk92iDIAw', 0.001),
 ('FOBRPlBHa3WPHFB5qYDlVg', 0.001),
 ('zZUnPeh2hEp0WydbAZEOOg', 0.001),
 ('QaELAmRcDc5TfJEylaaP8g', 0.001),
 ('xvu8G900tezTzbbfqmTKvA', 0.001),
 ('z5_82komKV3mI4ASGe2-FQ', 0.001),
 ('ttumcu6hWshk_EJVWrduDg', 0.001),
 ('f4_MRNHvN-yRn7EA8YWRxg', 0.001),
 ('UYACF30806j2mfbB5vdmJA', 0.001),
 ('QG13XBbgHWydzThRBGJtyw', 0.001),
 ('f6YuZP6iennHFVlnFJOXLQ', 0.001),
 ('I_6wY8_RsewziNnKhGZg4g', 0.001),
 ('q-v8elVPvKz0KvK69QSj1Q', 0.001),
 ('HwPGLzF_uXB3MF8bc5u5dg', 0.001),
 ('y4UuVowA9i3zj2hHyRMfHw', 0.001),
 ('1WBxJ2r3A2QYfRSEzgcmkQ', 0.001),
 ('-TT5e-YQU9xLb1JAGCGkQw', 0.001),
 ('6bbHSJ0PrgSxh7e5nigKMw', 0.001),
 ('4VmuXuSRhv5UxYUy3tMpiQ', 0.001),
 ('pVU2DdtBFppBAX5G5t3rcw', 0.001),
 ('q6QdLxYENKbVSo29ED2-JA', 0.001),
 ('cWw606W2PmcjTZrwLq8Hmg', 0.001),
 ('e9nlgxYgKrhE4VkwOo0LyQ', 0.001),
 ('Fta-vmTJYKGeVyyhm9D6Vw', 0.001),
 ('PamWn4FZwRiCYMdTw4lhcw', 0.001),
 ('qmcTQ4RSOnKqW5eqc1CEfw', 0.001),
 ('3o09Ya5VcRVMN3b2szZvkQ', 0.001),
 ('2LbG-TMDDk3Dxt18bfOdnQ', 

In [36]:
ranks = votes.reduceByKey(lambda x, y: x + y).sortBy(lambda x: x[1], ascending = False)
ranks.collect()

[('dIIKEfOgo0KqUfGQvGikPg', 6.141770837343267e-07),
 ('rt3PC7WCgCKsoufmQJELfw', 4.498681226296019e-07),
 ('KTU5QABqPrhvWmZPcqVojA', 4.2960232853667167e-07),
 ('NMmu4oC-YRqdi0WY9pV78g', 3.8085732673396424e-07),
 ('Zr4N9sG31fybEa-l3a_46Q', 3.73482586894213e-07),
 ('UmXzwFM2OkZHWJ5cJZlqkg', 3.58571967870274e-07),
 ('n5wQ7qm2QZY9iB8c25pzxA', 3.488759265665018e-07),
 ('HtCPdVujYJAZF4FKr-C9Lg', 3.4532057457318374e-07),
 ('JePOtJA36juFJv3LdkRvIQ', 3.303859741544938e-07),
 ('US4cngzcLZBPSgHVn9gq0A', 3.136410964400835e-07),
 ('VeYSD3Xgxd-R4yowet51tA', 3.1003003941552637e-07),
 ('ejU4_3dGN5QLEZxfU5UuRg', 3.0902439815203105e-07),
 ('JnPIjvC0cmooNDfsa9BmXg', 2.868847607132016e-07),
 ('jjHcf6XWMU7LmOEeUc9dqA', 2.8468621041504033e-07),
 ('YUxH_F9jMGAe7PIe8DO34Q', 2.821528917116168e-07),
 ('hTEEHPvMbAN6JCEUZhDQhg', 2.642926019633356e-07),
 ('MklqyU8t3AfR7V_9VWPGlA', 2.636788318633269e-07),
 ('yPLi_dtwzBk92m8Pqmxylw', 2.6318611412059256e-07),
 ('avRb9_uqiqFX71gzDL7ElA', 2.6318611412059256e-07),
 ('5Bh

In [33]:
N = links.count()
ranks = links.map(lambda line: (line[0], 1/N))
ranks.take(N)
sum = 1

while sum > 0.1:
    old_ranks = ranks
    votes = ranks.join(links) \
            .flatMap(calculateVotes)
    ranks = votes.reduceByKey(lambda x, y: x + y)
    errors = old_ranks.join(ranks).mapValues(lambda v: abs(v[0] - v[1]))
    sum = errors.values().sum()
    print(sum)
ranks.collect()

0.6122776780330376
0.7095322751700202
0.01268081238529119


[('4uSjk1cryPttErTp6zwa3Q', 5.288925123143311e-08),
 ('D4m0bLTkwZA3NpEjFhqULg', 6.200902279344555e-08),
 ('JVbqMyicxGWlEVPpyo6gpQ', 4.291930826832798e-08),
 ('LBsGQigIVfsvv-_GBsuKoQ', 1.582308788605815e-08),
 ('3qdOo7aWhIHWrKuT9jw3Gg', 4.8693248019213974e-08),
 ('HdUpeg9oYzvzHIgeJuM2uQ', 4.5567051439116445e-08),
 ('5D8fu3zjixrd6V89csWfvw', 2.483598814186634e-08),
 ('RSOfGM65pJiJiUmNjwDhHQ', 1.8219173741350574e-08),
 ('gbsm4QKtaSOVSDkVQ8s-hQ', 4.323475419182383e-09),
 ('_9WN_qmjbbRu6eFCMVXokw', 7.155270598670885e-09),
 ('nuxfSzP9sfVL9Mjva-5JsQ', 4.323475419182383e-09),
 ('-B-QEUESGWHPE_889WJaeg', 5.4403901057420336e-09),
 ('VQHljdk9KKpuAn2kw7GT3g', 4.323475419182383e-09),
 ('nUN2FOWj4G3ZmyCeDkq-FA', 1.887597853333011e-08),
 ('QtW6CE2L8yY3CdA_OWShRA', 4.323475419182383e-09),
 ('ZWD8UH1T7QXQr0Eq-mcWYg', 1.6854679521647452e-08),
 ('0xRAAStEi_dBFLD-1xSMHQ', 2.5017497867351323e-08),
 ('C3aIEUOx-vrHAeGVDnMAjg', 9.218857937869194e-09),
 ('KvbJLNmU1Noo5_6K1ZN0WQ', 4.323475419182383e-09),
 ('8Qv

The most popular user ranked by number of friends is dIIKEfOgo0KqUfGQvGikPg with a rank of 6.141770837343267e-07.